mllib/src/main/scala/org/apache/spark/mllib/clustering/PowerIterationClustering.scala

```diff
@@ -0,0 +1,220 @@
+   * @param nRuns  Number of runs for the KMeans clustering
+   * @return Tuple of (Seq[(Cluster Id,Cluster Center)],
+   *         Seq[(VertexId, ClusterID Membership)]
+   */
+  def run(sc: SparkContext,
+          G: Graph[Double, Double],
+          nClusters: Int,
+          nIterations: Int = defaultIterations,
+          nRuns: Int = defaultKMeansRuns)
+  : (Seq[(Int, Vector)], Seq[((VertexId, Vector), Int)]) = {
+    val (gUpdated, lambda, vt) = getPrincipalEigen(sc, G, nIterations)
+    // TODO: avoid local collect and then sc.parallelize.
+    val localVt = vt.collect.sortBy(_._1)
+    val vectRdd = sc.parallelize(localVt.map(v => (v._1, Vectors.dense(v._2))))
+    vectRdd.cache()
+    val model = KMeans.train(vectRdd.map {
+      _._2
+    }, nClusters, nRuns)
+    vectRdd.unpersist()
+    if (logger.isDebugEnabled) {
+      logger.debug(s"Eigenvalue = $lambda EigenVector: ${localVt.mkString(",")}")
+    }
+    val estimates = vectRdd.zip(model.predict(vectRdd.map(_._2)))
+    if (logger.isDebugEnabled) {
+      logger.debug(s"lambda=$lambda  eigen=${localVt.mkString(",")}")
+    }
+    val ccs = (0 until model.clusterCenters.length).zip(model.clusterCenters)
+    if (logger.isDebugEnabled) {
+      logger.debug(s"Kmeans model cluster centers: ${ccs.mkString(",")}")
+    }
+    val estCollected = estimates.collect.sortBy(_._1._1)
+    if (logger.isDebugEnabled) {
+      val clusters = estCollected.map(_._2)
+      val counts = estCollected.groupBy(_._2).mapValues {
+        _.length
+      }
+      logger.debug(s"Cluster counts: Counts: ${counts.mkString(",")}"
+        + s"\nCluster Estimates: ${estCollected.mkString(",")}")
+    }
+    (ccs, estCollected)
+  }
+
+
+  /**
+   * Create a Graph given an initial Vt0 and a set of Edges that
+   * represent the Normalized Affinity Matrix (W)
+   */
+  def createGraphFromEdges(sc: SparkContext,
+                           edgesRdd: RDD[Edge[Double]],
+                           nPoints: Int,
+                           optInitialVt: Option[Seq[(VertexId, Double)]] = None) = {
+
+    assert(nPoints > 0, "Must provide number of points from the original dataset")
+    val G = if (optInitialVt.isDefined) {
+      val initialVt = optInitialVt.get
+      val vertsRdd = sc.parallelize(initialVt)
+      Graph(vertsRdd, edgesRdd)
+    } else {
+      Graph.fromEdges(edgesRdd, -1.0)
+    }
+    G
+
+  }
+
+  /**
+   * Calculate the dominant Eigenvalue and Eigenvector for a given sparse graph
+   * using the PIC method
+   * @param sc
+   * @param G  Input Graph representing the Normalized Affinity Matrix (W)
+   * @param nIterations Number of iterations of the PIC algorithm
+   * @param optMinNormChange Minimum norm acceleration for detecting convergence
+   *                         - indicated as "epsilon" in the PIC paper
+   * @return
+   */
+  def getPrincipalEigen(sc: SparkContext,
+                        G: DGraph,
```

Comment:
actually I do not detect indentation issues (and neither does IJ nor the scala style checker.) can you point out?